src/Guide/Api/Utils.hs

```diff
@@ -89,3 +102,47 @@ instance (HasSwagger api, KnownSymbol name, KnownSymbol desc) =>
              symbolVal (Proxy @desc))
             Nothing
      in toSwagger (Proxy @api) & applyTags [tag]
+
+-- | Servant request details, can be captured by adding `RequestDetails :>` to an API branch.
+data RequestDetails = RequestDetails
+  { rdIp        :: Maybe IP  -- request ip address
+  , rdReferer   :: Maybe Text  -- request referrer
+  , rdUserAgent :: Maybe Text  -- request User-Agent
+  } deriving (Show, Generic)
+
+instance FromHttpApiData IP where
+  parseUrlPiece text
+    | Just ip <- readMaybe $ toString text = Right ip
+    | otherwise = Left "Not a parseable IP address"
+
+instance (HasSwagger api) => HasSwagger (RequestDetails :> api) where
+  toSwagger _ = toSwagger (Proxy :: Proxy api)
+
+instance (HasServer api context)
+  => HasServer (RequestDetails :> api) context where
+
+  type ServerT (RequestDetails :> api) m = RequestDetails -> ServerT api m
+
+  hoistServerWithContext _ pc nt s = hoistServerWithContext (Proxy :: Proxy api) pc nt . s
+
+  route Proxy context subserver = route (Proxy :: Proxy api) context $
+      subserver `addHeaderCheck` withRequest getRequestDetails
+    where
+      getRequestDetails :: Request -> DelayedIO RequestDetails
+      getRequestDetails req = pure $ RequestDetails ip referer useragent
+        where
+          ip :: Maybe IP
+          ip = maybe ipFromHost Just $ (dropEither "Forwarded-For") <|> (dropEither "X-Forwarded-For")
+          referer :: Maybe Text
+          referer = dropEither hReferer
+          useragent :: Maybe Text
+          useragent = dropEither hUserAgent
+
+          dropEither :: FromHttpApiData a => NHTH.HeaderName -> Maybe a
```

Comment:
Why is it called `dropEither`? It gets a header; `getHeader` is a better name. (Or pick something else.)